include/cppkafka/clonable_ptr.h

```diff
@@ -98,13 +96,41 @@ class ClonablePtr {
         return handle_.release();
     }
     
+    /**
+     * \brief Reset the internal pointer to a new one
+     */
+    void reset(T* ptr) {
+        handle_.reset(ptr);
+    }
+    
+    /**
+     * \brief Get the deleter
+     */
+    const Deleter& get_deleter() const {
+        return handle_.get_deleter();
+    }
+    
+    /**
+     * \brief Get the cloner
+     */
+    const Cloner& get_cloner() const {
+        return cloner_;
+    }
+    
     /**
      * \brief Indicates whether this ClonablePtr instance is valid (not null)
      */
     explicit operator bool() const {
         return static_cast<bool>(handle_);
     }
 private:
+    /**
+     * \brief Clones the internal pointer using the specified cloner function.
+     */
+    T* clone() const {
+        return cloner_ ? cloner_(get()) : get();
```

Comment:
If there's no cloner, this should probably throw, right? Otherwise you think you got a new pointer but in fact it's the same as the old one and you'll end up deleting it twice.

Reply:
Well this function is now private so its only purpose is to greatly simplify the constructor and copy constructor code  (see how it was written originally). If it was public I would agree with you.

Reply:
We can call it something else than `clone()` if you prefer, maybe `copy` or `mirror`?

Reply:
I don't think it's a matter of naming. The only case where you won't have a cloner function is if the `ClonablePtr` was moved, right? In that case, you won't have a pointer either. So this boils down to how you want to handle copy constructing/assigning from an empty `ClonablePtr`. I don't imagine all rdkafka destroy functions will work correctly if you pass in a null pointer, right? In which case, this shouldn't actually be returning just `get()` as that would end up blowing up later, given `get()` is a null pointer.

The result of copy constructing from an empty `ClonablePtr` should be that you have an empty pointer with an empty deleter (which is actually the same state `rhs` is in). A combination of pointer + non null deleter will probably go :boom:.

Reply:
Yeah well the times when you have a null cloner function is when 1) the `ClonablePtr` was moved and 2) the `ClonablePtr` was null to begin with - aka built via default `HeaderList` constructor. Whenever I use directly the rdkafka destroy or copy methods (which require a valid handle!) I always check if the `HeaderList` object is set (i.e. via bool operator). If not I never call the rdkafka functions. The only time when you can pass a null header handle is in the HEADERS_V macro inside `Producer::produce` because from what I can see i the rdkafka code, the headers are always checked before they are dereferenced, so in this case you can pass a null ptr. 
The combo of non-null pointer with null deleter can't exist...not sure where you see that?

Reply:
This is a sample code from rdkafka:
```
void rd_kafka_headers_destroy (rd_kafka_headers_t *hdrs) {
        rd_list_destroy(&hdrs->rkhdrs_list); <---- BOOM
        rd_free(hdrs);
}
```
I personally would have put a guard around this logic, just like calling `delete nullptr` doesn't blow up.

Reply:
Well, we can't change rdkafka's code so we have to deal with it. As it is now, if you construct a `ClonablePtr` using an empty cloner and then copy construct from it, it blows up. This shouldn't blow up and instead should just throw.

Reply:
I'm not sure if we're out of sync here...the only place where the cloner is called is inside the private method `ClonablePtr::try_clone()` (I renamed it since it does not always clone) and the validity of it is checked before being invoked. In the case of the copy constructor with an empty cloner (this is the standard non-owning case) it works perfectly fine, the handle is just copied as-is along with the DummyDeleter function. Don't think it will blow up. Unless I'm missing something...

Reply:
Well, it will blow up if the deleter is not dummy. This is used internally though so it should be fine.